src/api/nbt_template.py

```diff
@@ -0,0 +1,333 @@
+    """
+
+    __slots__ = ("tag_type", "value")
+
+    def __init__(self, tag_type: str, value: Any):
+        self.tag_type = tag_type
+        self.value = value
+
+    def __eq__(self, other: NBTEntry) -> bool:
+        return self.tag_type == other.tag_type and self.value == other.value
+
+    def __ne__(self, other: NBTEntry) -> bool:
+        return not self == other
+
+
+class NBTListEntry(UserList):
+    """
+    Container to represent a NBT List entry
+    """
+
+    tag_type = "list"
+
+    @property
+    def value(self):
+        return self.data
+
+
+class NBTCompoundEntry(UserDict):
+    """
+    Container to represent a NBT Compound entry
+    """
+
+    tag_type = "compound"
+
+    @property
+    def value(self):
+        return self.data
+
+
+class TemplateLoader:
+    """
+    Handles the loading of NBT template json files
+    """
+
+    def __init__(self, template_dir: str):
+        self._template_dir = template_dir
+
+        iter_chain = chain(
+            iglob(join(self._template_dir, "entities", "*.json")),
+            iglob(join(self._template_dir, "blockentities", "*.json")),
+            iglob(join(self._template_dir, "misc", "*.json")),
+        )
+        self._templates = {basename(p)[:-5]: p for p in iter_chain}
+        self._entity_template = self.load_template("entity")
+        self._item_template = self.load_template("item")
+
+    @property
+    def template_dir(self) -> str:
+        return self._template_dir
+
+    @property
+    def templates(self) -> Dict[str, str]:
+        return self._templates
+
+    def _tag_hook(self, json_object: dict) -> Union[dict, NBTStruct]:
+        if "tag_type" not in json_object:
+            return json_object
+
+        tag_type: str = json_object["tag_type"]
+
+        if tag_type.startswith("<") and tag_type.endswith(">"):
+            return self.load_template(tag_type[1:-1])
+
+        if tag_type == "compound":
+            structure = json_object["structure"]
+            if isinstance(structure, dict):
+                return NBTCompoundStructure(
+                    structure, default_value=json_object.get("default")
+                )
+            return NBTCompoundStructure(NBTStructure("any"))
+
+        if tag_type == "list":
+            return NBTListStructure(
+                json_object["schema"], default_value=json_object.get("default")
+            )
+
+        return NBTStructure(tag_type, default_value=json_object.get("default"))
+
+    def load_template(self, template_name: str) -> NBTCompoundStructure:
+        """
+        Loads the specified named NBT template file
+
+        :param template_name: The (normalized) name of the template file (IE: "creeper" for "creeper.json")
+        :return: The NBT template structure from the specified template file
+        """
+        if template_name not in self._templates:
+            raise FileNotFoundError()
+
+        if template_name == "entity" and hasattr(self, "_entity_template"):
```

Comment:
Why do we need "entity" and "item" as special cases? And why do we need to save them into variables in init?

Reply:
It was to save on the amount of times they were read from disc, but I guess it's not necessarily a good design and could be removed